Main.hs

```diff
@@ -3,35 +3,92 @@ module Main where
 import Control.Monad (unless)
 import Text.Printf (printf)
 
-data Expr
+data Expr 
+  = Num Double
+  | Sqrt Expr 
+  | Add Expr Expr       
+  | Sub Expr Expr       
+  | Mul Expr Expr       
+  | Div Expr Expr       
+  | Pow Expr Expr                
 
 instance Show Expr where
-  show = undefined
+  show (Num a) = show a
+  show (Sqrt a) = printf "sqrt(%s)" (show a)
+  show (Add a b) = printf "(%s + %s)" (show a) (show b)
+  show (Sub a b) = printf "(%s - %s)" (show a) (show b)
+  show (Mul a b) = printf "(%s * %s)" (show a) (show b)
+  show (Div a b) = printf "(%s / %s)" (show a) (show b)
+  show (Pow a b) = printf "(%s ^ %s)" (show a) (show b)
 
 instance Eq Expr where
-  (==) = undefined
+  Num a == Num b = a == b
+  Sqrt a == Sqrt b = a == b
+  Add a b == Add c d = a == c && b == d
+  Sub a b == Sub c d = a == c && b == d
+  Mul a b == Mul c d = a == c && b == d
+  Div a b == Div c d = a == c && b == d
+  Pow a b == Pow c d = a == c && b == d
+  _ == _ = False 
 
-data Error
+data Error = DivByZero Expr Expr
+            | NegativeSqrt Expr
 
 instance Show Error where
-  show = undefined
+  show (DivByZero a b ) = printf "Cannot divide %s by %s" (show a) (show b)
+  show (NegativeSqrt a ) = printf "Cannot take the square root of a negative number: %s" (show a)
 
 instance Eq Error where
-  (==) = undefined
+  DivByZero _ _ == DivByZero _ _ = True
+  NegativeSqrt _ == NegativeSqrt _ = True
+  _ == _ = False
```

Comment:
Whenever you have the default instance of a derivable type class, just derive it.